File: src/services/leadService.ts
import { v4 as uuidv4 } from 'uuid';
import { ILead, Lead } from '../models/Lead';
import { NormalizedLeadInput } from '../utils/leadNormalization';

export interface UpsertContext {
  p_id: string;
  source_id: string;
  rawRow: any;
}

export interface UpsertResult {
  action: 'inserted' | 'merged';
  lead: ILead;
}

/**
 * Find existing lead by matching any identifier (email, phone, aadhar)
 * Implements: "Duplicate if any identifier matches (email OR phone OR aadhar)"
 */
export const findExistingLead = async (
  normalized: NormalizedLeadInput
): Promise<ILead | null> => {
  const conditions: any[] = [];
  
  if (normalized.email) {
    conditions.push({ email: normalized.email });
  }
  
  if (normalized.phone_number) {
    conditions.push({ phone_number: normalized.phone_number });
  }
  
  if (normalized.aadhar_number) {
    conditions.push({ aadhar_number: normalized.aadhar_number });
  }
  
  if (conditions.length === 0) {
    return null;
  }
  
  // Find lead matching ANY of the identifiers
  return Lead.findOne({ $or: conditions });
};

/**
 * Merge strategy: Prefer existing non-empty values, fill gaps with incoming data
 * Deterministic precedence: Existing DB values take priority
 */
const mergeLeadData = (
  existing: ILead,
  incoming: NormalizedLeadInput,
  ctx: UpsertContext
): ILead => {
  // Fill missing fields with incoming data
  if (!existing.name && incoming.name) {
    existing.name = incoming.name;
  }
  
  if (!existing.phone_number && incoming.phone_number) {
    existing.phone_number = incoming.phone_number;
  }
  
  if (!existing.email && incoming.email) {
    existing.email = incoming.email;
  }
  
  if (!existing.aadhar_number && incoming.aadhar_number) {
    existing.aadhar_number = incoming.aadhar_number;
  }
  
  // Track source and product history
  if (!existing.sources_seen.includes(ctx.source_id)) {
    existing.sources_seen.push(ctx.source_id);
  }
  
  if (!existing.products_seen.includes(ctx.p_id)) {
    existing.products_seen.push(ctx.p_id);
  }
  
  // Store raw row for audit trail
  existing.merged_from.push({
    timestamp: new Date(),
    source_id: ctx.source_id,
    p_id: ctx.p_id,
    data: ctx.rawRow
  });
  
  return existing;
};

/**
 * Upsert lead: Insert new or merge with existing duplicate
 */
export const upsertLead = async (
  normalized: NormalizedLeadInput,
  ctx: UpsertContext
): Promise<UpsertResult> => {
  // Check for existing lead
  const existingLead = await findExistingLead(normalized);
  
  if (existingLead) {
    // Merge with existing lead
    const mergedLead = mergeLeadData(existingLead, normalized, ctx);
    await mergedLead.save();
    
    return {
      action: 'merged',
      lead: mergedLead
    };
  }
  
  // Create new lead
  const newLead = new Lead({
    lead_id: uuidv4(),
    name: normalized.name,
    phone_number: normalized.phone_number,
    email: normalized.email,
    aadhar_number: normalized.aadhar_number,
    source_id: ctx.source_id,
    p_id: ctx.p_id,
    created_at: new Date(),
    merged_from: [{
      timestamp: new Date(),
      source_id: ctx.source_id,
      p_id: ctx.p_id,
      data: ctx.rawRow
    }],
    sources_seen: [ctx.source_id],
    products_seen: [ctx.p_id],
    lead_score: null,
    score_reason: null
  });
  
  await newLead.save();
  
  return {
    action: 'inserted',
    lead: newLead
  };
};

File: bank-leads-java/bank-leads-backend/src/main/java/com/bankleads/bank_leads_backend/service/LeadService.java
package com.bankleads.bank_leads_backend.service;

import com.bankleads.bank_leads_backend.model.Lead;
import com.bankleads.bank_leads_backend.repository.LeadRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class LeadService {
    
    private final LeadRepository leadRepository;
    
    public Optional<Lead> findByLeadId(String leadId) {
        return leadRepository.findByLeadId(leadId);
    }
    
    @Transactional
    public UpsertResult upsertLead(Map<String, String> normalized, UpsertContext ctx) {
        // Check for existing lead
        Optional<Lead> existingOpt = findExistingLead(normalized);
        
        if (existingOpt.isPresent()) {
            Lead existing = existingOpt.get();
            Lead merged = mergeLeadData(existing, normalized, ctx);
            Lead saved = leadRepository.save(merged);
            return new UpsertResult("merged", saved);
        }
        
        // Create new lead
        Lead newLead = Lead.builder()
                .leadId(UUID.randomUUID().toString())
                .name(normalized.get("name"))
                .phoneNumber(normalized.get("phone_number"))
                .email(normalized.get("email"))
                .aadharNumber(normalized.get("aadhar_number"))
                .sourceId(ctx.sourceId)
                .pId(ctx.pId)
                .createdAt(LocalDateTime.now())
                .mergedFrom(new ArrayList<>())
                .sourcesSeen(new ArrayList<>(Collections.singletonList(ctx.sourceId)))
                .productsSeen(new ArrayList<>(Collections.singletonList(ctx.pId)))
                .leadScore(null)
                .scoreReason(null)
                .build();
        
        Lead.MergeRecord initialMerge = Lead.MergeRecord.builder()
                .timestamp(LocalDateTime.now())
                .sourceId(ctx.sourceId)
                .pId(ctx.pId)
                .data(ctx.rawRow)
                .build();
        
        newLead.getMergedFrom().add(initialMerge);
        
        Lead saved = leadRepository.save(newLead);
        return new UpsertResult("inserted", saved);
    }
    
    private Optional<Lead> findExistingLead(Map<String, String> normalized) {
        String email = normalized.get("email");
        String phone = normalized.get("phone_number");
        String aadhar = normalized.get("aadhar_number");
        
        // Try email first
        if (email != null && !email.isEmpty()) {
            List<Lead> leads = leadRepository.findByEmail(email);
            if (!leads.isEmpty()) {
                return Optional.of(leads.get(0));
            }
        }
        
        // Try phone
        if (phone != null && !phone.isEmpty()) {
            List<Lead> leads = leadRepository.findByPhoneNumber(phone);
            if (!leads.isEmpty()) {
                return Optional.of(leads.get(0));
            }
        }
        
        // Try aadhar
        if (aadhar != null && !aadhar.isEmpty()) {
            List<Lead> leads = leadRepository.findByAadharNumber(aadhar);
            if (!leads.isEmpty()) {
                return Optional.of(leads.get(0));
            }
        }
        
        return Optional.empty();
    }
    
    private Lead mergeLeadData(Lead existing, Map<String, String> incoming, UpsertContext ctx) {
        // Fill missing fields
        if ((existing.getName() == null || existing.getName().isEmpty()) && incoming.get("name") != null) {
            existing.setName(incoming.get("name"));
        }
        if ((existing.getPhoneNumber() == null || existing.getPhoneNumber().isEmpty()) && incoming.get("phone_number") != null) {
            existing.setPhoneNumber(incoming.get("phone_number"));
        }
        if ((existing.getEmail() == null || existing.getEmail().isEmpty()) && incoming.get("email") != null) {
            existing.setEmail(incoming.get("email"));
        }
        if ((existing.getAadharNumber() == null || existing.getAadharNumber().isEmpty()) && incoming.get("aadhar_number") != null) {
            existing.setAadharNumber(incoming.get("aadhar_number"));
        }
        
        // Track source and product history
        if (!existing.getSourcesSeen().contains(ctx.sourceId)) {
            existing.getSourcesSeen().add(ctx.sourceId);
        }
        if (!existing.getProductsSeen().contains(ctx.pId)) {
            existing.getProductsSeen().add(ctx.pId);
        }
        
        // Store raw row for audit trail
        Lead.MergeRecord mergeRecord = Lead.MergeRecord.builder()
                .timestamp(LocalDateTime.now())
                .sourceId(ctx.sourceId)
                .pId(ctx.pId)
                .data(ctx.rawRow)
                .build();
        
        existing.getMergedFrom().add(mergeRecord);
        
        return existing;
    }
    
    public static class UpsertResult {
        private final String action; // "inserted" or "merged"
        private final Lead lead;
        
        public UpsertResult(String action, Lead lead) {
            this.action = action;
            this.lead = lead;
        }
        
        public String getAction() { return action; }
        public Lead getLead() { return lead; }
    }
    
    public static class UpsertContext {
        private final String pId;
        private final String sourceId;
        private final Object rawRow;
        
        public UpsertContext(String pId, String sourceId, Object rawRow) {
            this.pId = pId;
            this.sourceId = sourceId;
            this.rawRow = rawRow;
        }
        
        public String getPId() { return pId; }
        public String getSourceId() { return sourceId; }
        public Object getRawRow() { return rawRow; }
    }
}
